### PR TITLE
Chef 16 add provides to resources

### PR DESCRIPTION
**Pull request by TheREK3R**
# Description

This adds 'provides' to all resources in this cookbook

## Issues Resolved

When testing another cookbook I had a resource error due to this behavior not being defined

## Check List

- [ X] All tests pass. See TESTING.md for details.
- [ NA] New functionality includes testing.
- [ NA069] New functionality has been documented in the README if applicable.

### CODE REVIEW

**Comment by kitchen-porter** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Error">
          1 Error
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">Please include a CHANGELOG entry.</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This Pull Request is probably missing tests.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: CHANGELOG.md [hunk 1]
```diff
@@ -2,6 +2,10 @@
 
 This file is used to list changes made in each version of the keepalived cookbook.
 
+## UNMERGED
```
Comment (ramereth) [inline CHANGELOG.md:5]:
```suggestion
## Unreleased
```

**Review by ramereth — CHANGES_REQUESTED** [review]

**Review by ramereth — APPROVED** [review]

**Comment by TheREK3R** [comment]
Actually this is my bad, I didnt realize we had pinned an old version of the cookbook, 

I do think this should be merged though. In the nginx cookbook we found that if another cookbook provides a similar resource defined with provides than it will supersede yours.

In that case the chef-nginx cookbook used `provides :nginx_site` while the sous-chefs cookbook did not. 

So although the sous-chefs was referenced first and chef_nginx was a dependency of a dependency the lack of provides made our resource be interpreted as `nginx_nginx_site` giving the 'nginx_site' provided by chef-nginx priority

 Thats obviously an edge case but IMO it should be defined.

**Comment by Xorima** [comment]
I will leave this to @tas50 and @lamont-granquist to advise on best practise here

**Comment by lamont-granquist** [comment]
I think you're mistake about that use case?  If you don't have a `provides` then `nginx/resources/nginx_site.rb` should only ever get automatically wired up as `nginx_nginx_site` and that has never changed.

I think the other way around though there was a problem, if you had a `nginx/resources/site.rb` without provides and `nginx/resources/nginx_site.rb` that did define `provides :nginx_site` then I think it is possible that pre-chef-16 without provides would always lose while post-chef-16 it now depends only on parse order.

( This kind of complexity gets at why I made everyone go through the pain in Chef 16 due to crazy edge conditions like this )

Going forwards though there's no difference in Chef-16, if you don't declare an explicit provides, then the implicit provides is the same "precedence level" so I don't think this matters either way.

At the same time I'm of the opinion that I like explicitness and self-documentation.  My encouragement is to define `provides` on every resource.

Tim disagrees, and I can't really cite any reason he's wrong.  I definitely have no plans of ever removing the implicit provides based on the cookbook+resource file.

Going forwards this may come down just to personal preference.

**Comment by TheREK3R** [comment]
Gotcha, thanks for that explanation! I'm glad it has switched in chef 16 to handle that. In that case please either close or merge this at your digression.

**Comment by ramereth** [comment]
I would feel better if we explicitly set the provides but I could go either way.

@tas50 @Xorima thoughts?

**Comment by Xorima** [comment]
If @lamont-granquist recommends it let's do it, at least this way it is the way chef are recommending at this point in time

**Comment by lamont-granquist** [comment]
Heh, well @tas50 recommends not using provides in this case.  I recommend using provides in this case.

So "Chef" is a bit split on the recommendation.

I would love to see sous-chefs play around with the problem and maybe figure out which of us is right and which is wrong (although I like the way Lance thinks...  =) ).

Kind of your chance to shine and figure out which direction we should go.

**Comment by lamont-granquist** [comment]
You should maybe ping Tim in the meeting this week and get his opinion.

**Comment by TheREK3R** [comment]
any updates? Should I close this?

**Comment by lamont-granquist** [comment]
No sous-chef meetings for the past two weeks, we keep having other stuff scheduled over it.

**Comment by TheREK3R** [comment]
Cool, thnaks or the feedback!

**Comment by ramereth** [comment]
I confirmed that this isn't needed. Thanks anyway!

**Pull request closed without merging**